libraries/chain/fork_database.cpp

```diff
@@ -603,15 +603,27 @@ namespace eosio::chain {
    }
 
    template<class BSP>
-   bool fork_database_t<BSP>::validated_block_exists(const block_id_type& id) const {
+   bool fork_database_t<BSP>::validated_block_exists(const block_id_type& id, const block_id_type& claimed_id) const {
       std::lock_guard g( my->mtx );
-      return my->validated_block_exists_impl(id);
+      return my->validated_block_exists_impl(id, claimed_id);
    }
 
+   // prerequisite: claimed_id is either id, or an ancestor of id
+   // returns true if block `id`, or one of its ancestors not older than claimed_id, is found in fork_db
+   // and `is_valid()`.
+   // ------------------------------------------------------------------------------------------------------
    template<class BSP>
-   bool fork_database_impl<BSP>::validated_block_exists_impl(const block_id_type& id) const {
-      auto itr = index.find( id );
-      return itr != index.end() && (*itr)->is_valid();
+   bool fork_database_impl<BSP>::validated_block_exists_impl(const block_id_type& id, const block_id_type& claimed_id) const {
+      bool id_present = false;
+
+      for (auto i = index.find(id); i != index.end(); i = index.find((*i)->previous())) {
+         id_present = true;
+         if ((*i)->is_valid())
+            return true;
+         if ((*i)->id() == claimed_id)
+            return false;
+      }
+      return id_present;
    }
```

Comment:
I think this should be something more like:
```cpp
   template<class BSP>
   bool fork_database_impl<BSP>::validated_block_exists_impl(const block_id_type& id, const block_id_type& claimed_id) const {
      assert(root);
      bool id_present = false;
      block_num_type claimed_block_num = block_header::num_from_id(claimed_id);

      for (auto i = index.find(id); i != index.end() && (*i)->block_num() >= claimed_block_num; i = index.find((*i)->previous())) {
         id_present = true;
         if ((*i)->id() == claimed_id)
            return (*i)->is_valid();
      }
      return id_present &&
             (root->id() == claimed_id || block_header::num_from_id(root->id()) > claimed_block_num);
   }

```

Reply:
What problem with my implementation does it solve?

The last return in your implementation already relies on the fact that `claimed_id` is an ancestor of `id`, since it compares block numbers.

So if we didn't find `claimed_id` while walking back up to toot, it means that `claimed_id` **has** to be either `root` or one of its ancestors, which are valid (they even are final).

Reply:
It solves the test cases I mentioned above that return the wrong result.

You can walk all the way back to root and not find `claimed_id` because the `claimed_id` is on a different branch than `id`. That is what the comparison of block numbers tells you.

Reply:
There is no testcase returning a wrong result, because  `claimed_id` is by precondition **not** on a different branch than `id`

Reply:
But the algorithm above also passes all the tests and returns `false` when not on the same branch so the precondition does not have to be true.

Reply:
In the use case we have for this function, the precondition has to be true.
I don't see why we have to make the function more complex to support a use case we don't need, and probably will never need.

Reply:
I don't think the algorithm above is more complex. I find it easier to reason that it does the right thing. However, I'll not press any more as your assert now verifies the precondition. You do have to remove the one test case now I would think.